### PR TITLE
Support overriding the partition table

### PR DESCRIPTION
**Pull request by floitsch**
## Summary by CodeRabbit

- **New Features**
	- Introduced options for custom partition tables and partitions during firmware extraction and flashing for ESP32 devices.
	- Enhanced `extract` and `flash` commands to support custom configurations.
	- Improved error handling with specific messages for invalid configurations.

- **Documentation**
	- Updated help messages for commands to reflect new options and usage.

### CODE REVIEW

File: tools/firmware.toit [hunk 1]
```diff
@@ -648,6 +652,14 @@ extract-cmd -> cli.Command:
         cli.OptionEnum "format" ["binary", "elf", "ubjson", "image", "qemu", "tar"]
             --help="Set the output format."
             --default="binary",
+        cli.Option "partition-table"
```
Comment (kasperl) [inline tools/firmware.toit:655]: partitions

File: tools/firmware.toit [hunk 2]
```diff
@@ -648,6 +652,14 @@ extract-cmd -> cli.Command:
...
+            --help="Override the partition table of the envelope."
+            --type="file",
+        cli.OptionPatterns "partition"
```
Comment (kasperl) [inline tools/firmware.toit:658]: Maybe change this to add-partition or something and keep partition as an alias?

File: tools/firmware.toit [hunk 3]
```diff
@@ -659,6 +671,14 @@ extract invocation/cli.Invocation -> none:
 
   config-path := invocation["config"]
 
+  partition-table-path/string? := invocation["partition-table"]
```
Comment (kasperl) [inline tools/firmware.toit:674]: partitions / {define,add}-partition

File: tools/firmware.toit [hunk 4]
```diff
@@ -834,33 +860,92 @@ extract-host invocation/cli.Invocation envelope/Envelope --config-encoded/ByteAr
...
+  ui := cli.ui
+
+  partition-table-path := invocation["partition-table"]
+
+  firmware-bin := extract-binary-esp32 envelope --config-encoded=config-encoded
+  binary := Esp32Binary firmware-bin
   flashing := envelope.entries.get AR-ENTRY-ESP32-FLASHING-JSON
       --if-present=: json.decode it
-      --if-absent=: throw "cannot create image without 'flashing.json'"
+      --if-absent=: ui.abort "Envelope is missing a 'flashing.json' entry"
```
Comment (kasperl) [inline tools/firmware.toit:884]:
```suggestion
      --if-absent=: ui.abort "Envelope is missing a 'flashing.json' entry."
```

File: tools/firmware.toit [hunk 5]
```diff
@@ -951,12 +1049,16 @@ flash-cmd -> cli.Command:
         cli.OptionInt "baud"
             --default=921600,
         cli.OptionEnum "chip" ["esp32", "esp32c3", "esp32s2", "esp32s3"]
+            --hidden
             --help="Deprecated. Don't use this option.",
         cli.OptionPatterns "partition"
```
Comment (kasperl) [inline tools/firmware.toit:1054]: add-partition or something...

File: tools/firmware.toit [hunk 6]
```diff
@@ -951,12 +1049,16 @@ flash-cmd -> cli.Command:
...
             ["file:<name>=<path>", "empty:<name>=<size>"]
             --help="Add a custom partition to the flashed image."
             --split-commas
             --multi,
+        cli.Option "partition-table"
```
Comment (kasperl) [inline tools/firmware.toit:1059]: partitions

File: tools/firmware.toit [hunk 7]
```diff
@@ -940,7 +1025,20 @@ esptool invocation/cli.Invocation -> none:
 
 flash-cmd -> cli.Command:
   return cli.Command "flash"
-      --help="Flash a firmware envelope to a device."
+      --help="""
+          Flash a firmware envelope to a device.
+
+          The partition table is extracted from the envelope unless the
+          --partition-table option is used.
```
Comment (kasperl) [inline tools/firmware.toit:1032]:
```suggestion
          '--partition-table' option is used.
```

**Review by kasperl — APPROVED** [review]
Looks good to me.

Not entirely sure about the --partition option. I think I'd prefer '--partition-add' or '--add-partition' so it is more cleanly different from --partitions.

File: tools/firmware.toit (hunk 1 above)
Comment (floitsch) [reply tools/firmware.toit]: done.

File: tools/firmware.toit (hunk 2 above)
Comment (floitsch) [reply tools/firmware.toit:656]: I will leave it as is for now. Will discuss this with you.

File: tools/firmware.toit (hunk 4 above)
Comment (floitsch) [reply tools/firmware.toit]: done.

File: tools/firmware.toit (hunk 7 above)
Comment (floitsch) [reply tools/firmware.toit]: done.

File: tools/firmware.toit (hunk 5 above)
Comment (floitsch) [reply tools/firmware.toit:1052]: leaving as is, for now.

File: tools/firmware.toit (hunk 6 above)
Comment (floitsch) [reply tools/firmware.toit]: done.

**Pull request merged**